### ISSUE TITLE
All my consul agents are crashing at start

### ISSUE THREAD

**Issue by kedare** [hashicorp/consul#9097]
#### Overview of the Issue

Today, when adding a new member on my consul cluster (agent), I got a panic instead of a working consul (the config has not been changed for months), I see that even restarting existing agents cause them to panic at start.
The existing running agents are working fine if not restarted.

#### Reproduction Steps

Steps to reproduce this issue, eg:

1. Install consul on a new server (in my case those are LXD containers)
1. Set the configuration (see later)
1. Start the service or `consul agent -data-dir=/var/lib/consul -config-dir=/etc/consul.d`

### Consul info for both Client and Server

<details>
  <summary>Client Configuration</summary>

```
log_level = "info"
datacenter = "frnte1"
addresses {
  http = "[::]"
  dns = "[::]"
}
bind_addr = "[::]" 
retry_join = ["frnte1-xxxx.xxxx.net"]
service {
  name = "node-exporter"
  tags = ["monitor"]
  port = 9100
  meta {
    metrics_path = "/metrics"
  }
}
```

</details>

<details>
  <summary>Server Configuration</summary>

```
log_level = "info"
bootstrap_expect = 1
server = true
ui = true
datacenter = "frnte1"
addresses {
  http = "[::]"
  dns = "[::]"
} 
service {
  name = "prometheus"
  tags = ["monitor"]
  port = 9090
  meta {
    metrics_path = "/metrics"
  }
}
service {
  name = "node-exporter"
  tags = ["monitor"]
  port = 9100
  meta {
    metrics_path = "/metrics"
  }
}
```
</details>

<details>
  <summary>Server info</summary>

```
agent:                                                                                         
        check_monitors = 0                                                                                                                                                                    
        check_ttls = 0                                                                                                                                                                        
        checks = 0                                                                                                                                                                            
        services = 2                                                                                                                                                                          
build:                                                                                                                                                                                        
        prerelease =                                                                                                                                                                          
        revision =                                                                                                                                                                            
        version = 1.5.2
consul:                  
        acl = disabled
        bootstrap = true
        known_datacenters = 1
        leader = true
        leader_addr = [2001:xxxx:xxxx:xxxx:xxxx:3eff:fe38:109b]:8300                             
        server = true                   
raft:                                                                                          
        applied_index = 1628773
        commit_index = 1628773                                                                 
        fsm_pending = 0                                                                        
        last_contact = 0    
        last_log_index = 1628773
        last_log_term = 20                                                                     
        last_snapshot_index = 1622441
        last_snapshot_term = 19
        latest_configuration = [{Suffrage:Voter ID:f7987789-7fe9-a898-d492-1ee8f679d274 Address:10.13.37.143:8300}]                                                                           
        latest_configuration_index = 1                                                         
        num_peers = 0
        protocol_version = 3
        protocol_version_max = 3
        protocol_version_min = 0
        snapshot_version_max = 1
        snapshot_version_min = 0
        state = Leader
        term = 20
runtime:
        arch = amd64
        cpu_count = 2
        goroutines = 154
        max_procs = 2
        os = linux
        version = go1.13.5
serf_lan:
        coordinate_resets = 0
        encrypted = false
        event_queue = 0
        event_time = 20
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 1
        member_time = 149
        members = 12
        query_queue = 0
        query_time = 1
serf_wan:
        coordinate_resets = 0
        encrypted = false
        event_queue = 0
        event_time = 1
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 1
        members = 1
        query_queue = 0
        query_time = 1
```

</details>

### Operating system and Environment details

Ubuntu 20.04, amd64, in LXD containers

#### Agent startup

```
$ consul agent -data-dir=/var/lib/consul -config-dir=/etc/consul.d                                                                                           [6/833]
==> Starting Consul agent...                                                                                                                                                                  
           Version: '1.5.2-dev'                                                                
           Node ID: 'dc6d0551-53c7-97a4-04e1-xxxx'           
         Node name: 'frnte1-xxxx'                                                        
        Datacenter: 'frnte1' (Segment: '')                                                     
            Server: false (Bootstrap: false)                                                   
       Client Addr: [127.0.0.1] (HTTP: 8500, HTTPS: -1, gRPC: -1, DNS: 8600)
      Cluster Addr: 2001:xxxx:xxxx:xxxx:xxxx:3eff:fe6d:fb68 (LAN: 8301, WAN: 8302)
           Encrypt: Gossip: false, TLS-Outgoing: false, TLS-Incoming: false, Auto-Encrypt-TLS: false

==> Log data will now stream in as it occurs:

    2020/11/04 09:31:47 [INFO] serf: EventMemberJoin: frnte1-xxxx 2001:xxxx:xxxx:xxxx:xxxx:3eff:fe6d:fb68
    2020/11/04 09:31:47 [WARN] serf: Failed to re-join any previously known node
    2020/11/04 09:31:47 [INFO] agent: Started DNS server [::]:8600 (udp)
    2020/11/04 09:31:47 [INFO] agent: Started DNS server [::]:8600 (tcp)
    2020/11/04 09:31:47 [INFO] agent: Started HTTP server on [::]:8500 (tcp)
    2020/11/04 09:31:47 [INFO] agent: started state syncer
==> Consul agent running!
    2020/11/04 09:31:47 [INFO] agent: Retry join LAN is supported for: aliyun aws gce mdns os packet
    2020/11/04 09:31:47 [WARN] manager: No servers available
    2020/11/04 09:31:47 [ERR] agent: failed to sync remote state: No known Consul servers
    2020/11/04 09:31:47 [INFO] agent: Joining LAN cluster...
    2020/11/04 09:31:47 [INFO] agent: (LAN) joining: [frnte1-xxxx.xxxx.net]
panic: runtime error: slice bounds out of range [:9] with length 8

 goroutine 99 [running]:
github.com/miekg/dns.ClientConfigFromReader(0x2a3bb00, 0xc000010028, 0x0, 0x0, 0xc000010028)
        github.com/miekg/dns/clientconfig.go:94 +0x8d1
github.com/miekg/dns.ClientConfigFromFile(0x265188e, 0x10, 0x0, 0x0, 0x0)
        github.com/miekg/dns/clientconfig.go:29 +0xde
github.com/hashicorp/memberlist.(*Memberlist).tcpLookupIP(0xc0003f0d80, 0xc00059e0c0, 0x1d, 0x206d, 0x0, 0x0, 0x0, 0x0, 0x100000000000054)
        github.com/hashicorp/memberlist/memberlist.go:285 +0xd4
github.com/hashicorp/memberlist.(*Memberlist).resolveAddr(0xc0003f0d80, 0xc00059e0c0, 0x22, 0x8, 0xc000530b10, 0x17c069b, 0x0, 0x0)
        github.com/hashicorp/memberlist/memberlist.go:350 +0x209
github.com/hashicorp/memberlist.(*Memberlist).Join(0xc0003f0d80, 0xc00054e120, 0x1, 0x1, 0x51, 0x0, 0x0)
        github.com/hashicorp/memberlist/memberlist.go:232 +0xc1
github.com/hashicorp/serf/serf.(*Serf).Join(0xc0003b7200, 0xc00054e120, 0x1, 0x1, 0x1, 0x0, 0x0, 0x0)
        github.com/hashicorp/serf/serf/serf.go:637 +0x113
github.com/hashicorp/consul/agent/consul.(*Client).JoinLAN(0xc0005a43f0, 0xc00054e120, 0x1, 0x1, 0x1, 0x1, 0xc000530d30)
        github.com/hashicorp/consul/agent/consul/client.go:237 +0x52
github.com/hashicorp/consul/agent.(*Agent).JoinLAN(0xc0000e4240, 0xc00054e120, 0x1, 0x1, 0x0, 0x0, 0x24)
        github.com/hashicorp/consul/agent/agent.go:1759 +0xff
github.com/hashicorp/consul/agent.(*retryJoiner).retryJoin(0xc000530f80, 0xc00048be80, 0x1e77f07)
        github.com/hashicorp/consul/agent/retry_join.go:124 +0x3c4
github.com/hashicorp/consul/agent.(*Agent).retryJoinLAN(0xc0000e4240)
        github.com/hashicorp/consul/agent/retry_join.go:23 +0xf3
created by github.com/hashicorp/consul/agent.(*Agent).Start
        github.com/hashicorp/consul/agent/agent.go:546 +0xc6b
```

**Comment by kedare** [comment]
I could solve the issue by completely wiping the cluster and starting from scratch but now it's happening again... :(

**Comment by dnephin** [comment]
Thank you for the bug report. It looks like this is running version 1.5.2, is that correct? Are you able to reproduce with a recent version of Consul (the latest in the 1.7.x, or 1.8.x series)?

**Comment by kedare** [comment]
Hello,
I will try with the last version.
Do I need to wipe the cluster to upgrade it ? Or just remplacing the consul binary should do it ?

**Comment by iforapsy** [comment]
We ran into this same problem with our Consul cluster.

Looking at the source code of `github.com/miekg/dns/clientconfig.go`, I learned that Consul parses `/etc/resolv.conf` at process startup. Due to a typo in the string indexing, it performs an out-of-bound memory read if there is a word of exactly 8 characters in the `options` line of that file. I got Consul to start up temporarily removing all such words.
```bash
sudo sed -i.BACKUP '/^option/s/[^ ]\{8\}//' /etc/resolv.conf
sudo systemctl start consul
sleep 5s
sudo mv /etc/resolv.conf.BACKUP /etc/resolv.conf
```
For our `resolv.conf` file, that changed the line "options edns0 trust-ad" to "options edns0 " (as "trust-ad" is 8 characters). This workaround works because Consul only reads the file once, at start.

Alternatively, you can fix the problem permanently by applying this patch and rebuilding Consul.
```
diff --git a/vendor/github.com/miekg/dns/clientconfig.go b/vendor/github.com/miekg/dns/clientconfig.go
index 0a1f5a92c..fc6a0f3f0 100644
--- a/vendor/github.com/miekg/dns/clientconfig.go
+++ b/vendor/github.com/miekg/dns/clientconfig.go
@@ -80,13 +80,13 @@ func ClientConfigFromFile(resolvconf string) (*ClientConfig, error) {
                                case len(s) >= 8 && s[:8] == "timeout:":
                                        n, _ := strconv.Atoi(s[8:])
                                        if n < 1 {
                                                n = 1
                                        }
                                        c.Timeout = n
-                               case len(s) >= 8 && s[:9] == "attempts:":
+                               case len(s) >= 8 && s[:8] == "attempts:":
                                        n, _ := strconv.Atoi(s[9:])
                                        if n < 1 {
                                                n = 1
                                        }
                                        c.Attempts = n
                                case s == "rotate":
```

We also checked Consul 1.8.4. That version already has this fix, and no changes to `/etc/resolv.conf` were necessary.

**Comment by dnephin** [comment]
@iforapsy Thank you for tracking down the problem and confirming it is fixed in recent versions. It looks like it was fixed in the library in https://github.com/miekg/dns/pull/642/files (the fix there used 9 for the indexes instead of 8).  Unfortunately I wasn't able to track down when we included that fix in Consul because it seems to pre-date our use of Go modules.

@kedare sorry I missed your question. Just updating the binary and restarting should be fine, no need to remove the data.

Since this issue has been fixed, I'm going to close it.
